File: BEmu/ReferenceDataRequest/ElementReference.cs
//------------------------------------------------------------------------------
// <copyright project="BEmu" file="ReferenceDataRequest/ElementReference.cs" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

namespace BEmu.ReferenceDataRequest
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Text;

    internal class ElementReference : Element
    {
        private readonly ElementReferenceSecurityDataArray _securities;

        internal ElementReference(MessageReference arg)
        {
            this._securities = new ElementReferenceSecurityDataArray((ElementReferenceSecurityDataArray)arg.Elements.First());
        }

        public override int NumElements { get { return 1; } }
        public override int NumValues { get { return 1; } }
        public override Name Name { get { return new Name("securityData"); } }
        public override IEnumerable<Element> Elements { get { yield return this._securities; } }
        public override bool IsNull { get { return false; } }
        public override bool IsArray { get { return false; } }
        public override bool IsComplexType { get { return true; } }

        public override Element this[string name]
        {
            get
            {
                if (name.ToUpper() == this._securities.Name.ToString().ToUpper())
                    return this._securities;
                else
                    return base[name];
            }
        }

        public override object this[string name, int index]
        {
            get
            {
                ElementReferenceSecurityDataArray elm = (ElementReferenceSecurityDataArray)this[name]; //checks name

                int elmIndex = 0;
                foreach (var item in elm.Elements)
                {
                    if (elmIndex == index)
                        return item;
                    elmIndex++;
                }

                return base[name, index];
            }
        }

        public override Element GetElement(string name)
        {
            return this[name];
        }

        internal override StringBuilder PrettyPrint(int tabIndent)
        {
            StringBuilder result = new StringBuilder();
            result.AppendFormat("ReferenceDataResponse (choice) = {{{0}", Environment.NewLine);
            result.Append(this._securities.PrettyPrint(1));
            result.Append("}" + Environment.NewLine);
            return result;
        }

    }
}

File: BEmuJava/bemu/src/main/java/com/bemu/BEmu/ReferenceDataRequest/ElementReference.java
package com.bemu.BEmu.ReferenceDataRequest;

import com.bemu.BEmu.Element;
import com.bemu.BEmu.Name;

public class ElementReference extends Element
{
    private final ElementReferenceSecurityDataArray _securities;

    ElementReference(MessageReference arg)
    {
        this._securities = new ElementReferenceSecurityDataArray((ElementReferenceSecurityDataArray)arg.elements().get(0));
    }
    
    public Name name()
    {
    	return new Name("securityData");
    }
    
    public int numValues()
    {
    	return 1;
    }
    
    public int numElements()
    {
    	return 1;
    }
    
    public boolean isComplexType()
    {
    	return true;
    }
    
    public boolean isArray()
    {
    	return false;
    }
    
    public boolean isNull()
    {
    	return false;
    }

    public Element getElement(String name) throws Exception
    {
        if (name.toUpperCase().equals(this._securities.name().toString().toUpperCase()))
            return this._securities;
        
        else
            return super.getElement(name);
    }
    
    protected StringBuilder prettyPrint(int tabIndent)
    {
        StringBuilder result = new StringBuilder();

        result.append(String.format("ReferenceDataResponse (choice) = {%s", System.getProperty("line.separator")));
        result.append(this._securities.prettyPrint(1));
        result.append(String.format("}%s", System.getProperty("line.separator")));
        
        return result;
    }
    
}
